File: SortVision/public/code/merge/csharp/mergeSort.cs
using System;

public static class MergeSort {
    // Time: Best/Avg/Worst O(n log n), Space: O(n)
    public static void Sort(int[] arr) {
        if (arr == null || arr.Length <= 1) return;
        SortRange(arr, 0, arr.Length - 1);
    }

    private static void SortRange(int[] arr, int left, int right) {
        if (left >= right) return;
        int mid = left + (right - left) / 2;
        SortRange(arr, left, mid);
        SortRange(arr, mid + 1, right);
        Merge(arr, left, mid, right);
    }

    private static void Merge(int[] arr, int left, int mid, int right) {
        int[] temp = new int[right - left + 1];
        int i = left, j = mid + 1, k = 0;

        while (i <= mid && j <= right) {
            temp[k++] = arr[i] <= arr[j] ? arr[i++] : arr[j++];
        }
        while (i <= mid) temp[k++] = arr[i++];
        while (j <= right) temp[k++] = arr[j++];

        for (int x = 0; x < temp.Length; x++) arr[left + x] = temp[x];
    }

    public static void Main() {
        int[] arr = {64, 25, 12, 22, 11};
        Sort(arr);
        Console.WriteLine(string.Join(", ", arr));
    }
}

File: SortVision/public/code/merge/cpp/mergeSort.cpp
#include <iostream>
#include <vector>
using namespace std;

// Merge Sort
// Time: Best/Avg/Worst O(n log n), Space: O(n)
static void mergeVec(vector<int>& arr, int l, int m, int r) {
    vector<int> left(arr.begin() + l, arr.begin() + m + 1);
    vector<int> right(arr.begin() + m + 1, arr.begin() + r + 1);

    int i = 0, j = 0, k = l;
    while (i < (int)left.size() && j < (int)right.size()) {
        arr[k++] = (left[i] <= right[j]) ? left[i++] : right[j++];
    }
    while (i < (int)left.size()) arr[k++] = left[i++];
    while (j < (int)right.size()) arr[k++] = right[j++];
}

void mergeSort(vector<int>& arr, int l, int r) {
    if (l >= r) return;
    int m = l + (r - l) / 2;
    mergeSort(arr, l, m);
    mergeSort(arr, m + 1, r);
    mergeVec(arr, l, m, r);
}

int main() {
    vector<int> arr = {64, 25, 12, 22, 11};
    mergeSort(arr, 0, (int)arr.size() - 1);
    for (int x : arr) cout << x << ' ';
    cout << '\n';
}
